tensorflow_probability/python/experimental/mcmc/covariance_reducer.py

```diff
@@ -0,0 +1,260 @@
+    selects among three different covariance computations:
+    - `event_ndims=0` treats the samples as a `[5, 7]` batch of scalar random
+      variables, and computes their variances in batch.  The shape of the result
+      is `[5, 7]`.
+    - `event_ndims=1` treats the samples as a `[5]` batch of vector random
+      variables of shape `[7]`, and computes their covariances in batch.  The
+      shape of the result is `[5, 7, 7]`.
+    - `event_ndims=2` treats the samples as a single random variable of
+      shape `[5, 7]` and computes its covariance.  The shape of the result
+      is `[5, 7, 5, 7]`.
+
+    For nested latent samples, both `event_ndims` and `ddof` must be either
+    a single value or an identical structure. For example, if the chain state
+    is the tuple of dictionaries ({pairA, pairB}, {pairC}), one could use
+    scalar values of `event_ndims` (integer or `None`) and `ddof` (integer) that
+    apply to all states in the latent. If not, `event_ndims` and `ddof` must
+    have structures that ressemble ({keyA: argA, keyB: argB}, {keyC: argC}).
+
+    Args:
+      event_ndims: A (possibly nested) structure of integers, or `None`. Defines
+        the number of inner-most dimensions that represent the event shape.
+        Specifying `None` returns all cross product terms (no batching)
+        and is the default.
+      ddof: A (possibly nested) structure of integers that represent the
+        requested dynamic degrees of freedom. For example, use `ddof=0`
+        for population covariance and `ddof=1` for sample covariance. Defaults
+        to the population covariance.
+      name: Python `str` name prefixed to Ops created by this function.
+        Default value: `None` (i.e., 'covariance_reducer').
+    """
+    self._parameters = dict(
+        event_ndims=event_ndims,
+        ddof=ddof,
+        name=name or 'covariance_reducer'
+    )
+
+  def initialize(self, initial_chain_state=None, initial_kernel_results=None):
+    """Initializes a `RunningCovarianceState` using previously defined metadata.
+
+    For calculation purposes, the `initial_chain_state` does not count as a
+    sample. This is a deliberate decision that ensures consistency across
+    sampling procedures (i.e. `tfp.mcmc.sample_chain` follows similar
+    semantics).
+
+    Args:
+      initial_chain_state: A (possibly nested) structure of `Tensor`s or Python
+        `list`s of `Tensor`s representing the current state(s) of the Markov
+        chain(s). It is used to infer the shape and dtype of future samples.
+      initial_kernel_results: A (possibly nested) `tuple`, `namedtuple` or
+        `list` of `Tensor`s representing internal calculations made in a related
+        `TransitionKernel`. For streaming covariance, this argument also has no
+        influence on computation; hence, it is also `None` by default.
+        Nonetheless, this argument is still accepted as it will be supplied by
```

Comment:
Sure, but that's somewhat backwards. The reason its here is because it's in the super-class and we're following that interface.